### PR TITLE
restruct blobaccess

### PR DESCRIPTION
**Pull request by mandelsoft**
## Description

- get rid of blobaccess package dependencies from blobaccess implementation packages.
- provide access package with access to all known implementations to simplify the usage for external users
- separate blobaccess for external users from blobaccessfor packages used by BlobAccess implementations to
  avoid cycles.

- separate deprecated functions and types into a `dedicated.go` file deprecated.go to prepare automated cleanup during migration.

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -688,6 +688,16 @@ github.com/mandelsoft/goutils v0.0.0-20240605092610-3d3ecdc682ea h1:VDau62ZodJev
...
+github.com/mandelsoft/goutils v0.0.0-20240607120900-c70e5479360b h1:qtJDWErrFlctzuord4yZq8u2EYD+ReXqAfykM5UDhSU=
+github.com/mandelsoft/goutils v0.0.0-20240607120900-c70e5479360b/go.mod h1:9TJgkwSY43RWHiIAAz7fL8SEIHf0L13Pk4w8fDIt+i4=
+github.com/mandelsoft/goutils v0.0.0-20240608121958-b990a46c6aa2 h1:EC3lToqJdkAmtbkrizqVxAPMgbXEoJqppkdKF53aIAg=
+github.com/mandelsoft/goutils v0.0.0-20240608121958-b990a46c6aa2/go.mod h1:9TJgkwSY43RWHiIAAz7fL8SEIHf0L13Pk4w8fDIt+i4=
+github.com/mandelsoft/goutils v0.0.0-20240608130105-a79e05d3ba99 h1:DIQZb85AkMjIA3dTzLnW0LzxYSMryPq+n0mOJSLai/s=
+github.com/mandelsoft/goutils v0.0.0-20240608130105-a79e05d3ba99/go.mod h1:9TJgkwSY43RWHiIAAz7fL8SEIHf0L13Pk4w8fDIt+i4=
+github.com/mandelsoft/goutils v0.0.0-20240608130415-ae767b734156 h1:LLvRlPQgpzGtWbMvcvFmWBdzgNOn/9kBmiDu24kGncw=
+github.com/mandelsoft/goutils v0.0.0-20240608130415-ae767b734156/go.mod h1:9TJgkwSY43RWHiIAAz7fL8SEIHf0L13Pk4w8fDIt+i4=
+github.com/mandelsoft/goutils v0.0.0-20240608132424-ec9fb7fa611a h1:7ZnEdaPeoshk2I1GfZcE51NkSjGwfuVoM2Jf/tAXB2w=
+github.com/mandelsoft/goutils v0.0.0-20240608132424-ec9fb7fa611a/go.mod h1:9TJgkwSY43RWHiIAAz7fL8SEIHf0L13Pk4w8fDIt+i4=
```
Comment (hilmarf) [inline go.sum:700]: how does this `go.sum` work? Sometimes there are two entries per dependency, sometimes more than 2...

File: pkg/blobaccess/file/access.go [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+	"github.com/mandelsoft/vfs/pkg/vfs"
+	"github.com/opencontainers/go-digest"
+
+	"github.com/open-component-model/ocm/pkg/blobaccess/bpi"
+	"github.com/open-component-model/ocm/pkg/iotools"
+	"github.com/open-component-model/ocm/pkg/utils"
+)
+
+type (
+	_nopCloser  = iotools.NopCloser
```
Comment (hilmarf) [inline pkg/blobaccess/file/access.go:17]: What's the benefit of redeclaring types? Doesn't it hide the usage of `iotools.NopCloser` and make refactorings more complex?

File: pkg/blobaccess/file/access.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+
+func (a *fileDataAccess) Reader() (io.ReadCloser, error) {
+	file, err := a.fs.Open(a.path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "file %q", a.path)
+	}
+	return file, nil
+}
+
+func (a *fileDataAccess) Validate() error {
```
Comment (hilmarf) [inline pkg/blobaccess/file/access.go:54]: Can we please have some short documentation? `Validate()` could mean a lot of different things in context of OCM... what's validated? Is the access correct? Does the files exist? Are the digests validated? ...?

File: pkg/blobaccess/file/access.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+	file       vfs.File
+	filesystem vfs.FileSystem
+}
+
+var (
+	_ bpi.BlobAccessBase = (*temporaryFileBlob)(nil)
+	_ bpi.FileLocation   = (*temporaryFileBlob)(nil)
+)
+
+func (b *temporaryFileBlob) Validate() error {
```
Comment (hilmarf) [inline pkg/blobaccess/file/access.go:186]: Documentation? This `Validate()` seem to work differently, since it uses locks...

File: pkg/blobaccess/file/access.go [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+func ForTemporaryFile(mime string, temp vfs.File, fss ...vfs.FileSystem) bpi.BlobAccess {
+	return bpi.NewBlobAccessForBase(&temporaryFileBlob{
+		_blobAccess: BlobAccess(mime, temp.Name(), fss...),
+		filesystem:  utils.FileSystem(fss...),
+		path:        temp.Name(),
+		file:        temp,
+	})
+}
+
+func ForTemporaryFileWithMeta(mime string, digest digest.Digest, size int64, temp vfs.File, fss ...vfs.FileSystem) bpi.BlobAccess {
```
Comment (hilmarf) [inline pkg/blobaccess/file/access.go:236]: What is `Meta`?

File: pkg/blobaccess/file/access.go [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+func ForTemporaryFilePath(mime string, temp string, fss ...vfs.FileSystem) bpi.BlobAccess {
+	return bpi.NewBlobAccessForBase(&temporaryFileBlob{
+		_blobAccess: BlobAccess(mime, temp, fss...),
+		filesystem:  utils.FileSystem(fss...),
+		path:        temp,
+	})
+}
+
+func ForTemporaryFilePathWithMeta(mime string, digest digest.Digest, size int64, temp string, fss ...vfs.FileSystem) bpi.BlobAccess {
```
Comment (hilmarf) [inline pkg/blobaccess/file/access.go:253]: What is `Meta`?
Can't we stick to the super fancy `OptionUtils` pattern?

File: pkg/blobaccess/file/access.go [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+////////////////////////////////////////////////////////////////////////////////
+
+// TempFile holds a temporary file that should be kept open.
+// Close should never be called directly.
+// It can be passed to another responsibility realm by calling Release
+// For example to be transformed into a TemporaryBlobAccess.
+// Close will close and remove an unreleased file and does
+// nothing if it has been released.
+// If it has been releases the new realm is responsible.
```
Comment (hilmarf) [inline pkg/blobaccess/file/access.go:269]:
```suggestion
// If it has been released the new realm is responsible.
```

File: pkg/blobaccess/file/access.go [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+	return t.temp.Name()
+}
+
+func (t *TempFile) FileSystem() vfs.FileSystem {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	return t.filesystem
+}
+
+func (t *TempFile) Release() vfs.File {
```
Comment (hilmarf) [inline pkg/blobaccess/file/access.go:301]: Documentation?

File: pkg/blobaccess/file/access.go [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	return t.temp.Sync()
+}
+
+func (t *TempFile) AsBlob(mime string) bpi.BlobAccess {
+	return ForTemporaryFile(mime, t.Release(), t.filesystem)
+}
+
+func (t *TempFile) Close() error {
```
Comment (hilmarf) [inline pkg/blobaccess/file/access.go:328]: Documentation?

File: pkg/blobaccess/forward.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+func DataAccessForBytes(data []byte, origin ...string) bpi.DataSource {
+	return standard.DataAccessForData(data, origin...)
+}
+
+// DataAccessForData wraps a bytes slice into a DataAccess.
+func DataAccessForData(data []byte, origin ...string) bpi.DataSource {
+	return standard.DataAccessForData(data, origin...)
+}
+
+func DataAccessForString(data string, origin ...string) bpi.DataSource {
```
Comment (hilmarf) [inline pkg/blobaccess/forward.go:33]:
```suggestion
func DataSourceForString(data string, origin ...string) bpi.DataSource {
```

File: pkg/blobaccess/forward.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+	return standard.DataAccessForString(data, origin...)
+}
+
+// ForString wraps a string into a BlobAccess, which does not need a close.
+func ForString(mime string, data string) BlobAccess {
```
Comment (hilmarf) [inline pkg/blobaccess/forward.go:38]:
```suggestion
func BlobAccessForString(mime string, data string) BlobAccess {
```

File: pkg/blobaccess/forward.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	return standard.ForString(mime, data)
+}
+
+func ProviderForString(mime, data string) BlobAccessProvider {
```
Comment (hilmarf) [inline pkg/blobaccess/forward.go:42]:
```suggestion
func BlobAccessProviderForString(mime, data string) BlobAccessProvider {
```

File: pkg/blobaccess/forward.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+	return standard.ProviderForString(mime, data)
+}
+
+// ForData wraps data into a BlobAccess, which does not need a close.
+func ForData(mime string, data []byte) BlobAccess {
```
Comment (hilmarf) [inline pkg/blobaccess/forward.go:47]:
```suggestion
func BlobAccessForBytes(mime string, data []byte) BlobAccess {
```

File: pkg/blobaccess/forward.go [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+	return standard.ForData(mime, data)
+}
+
+func ProviderForData(mime string, data []byte) BlobAccessProvider {
```
Comment (hilmarf) [inline pkg/blobaccess/forward.go:51]:
```suggestion
func BlobAccessProviderForBytes(mime string, data []byte) BlobAccessProvider {
```

File: pkg/blobaccess/forward.go [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+// TempFile holds a temporary file that should be kept open.
+// Close should never be called directly.
+// It can be passed to another responsibility realm by calling Release
+// For example to be transformed into a TemporaryBlobAccess.
+// Close will close and remove an unreleased file and does
+// nothing if it has been released.
+// If it has been releases the new realm is responsible.
+// to close and remove it.
+type TempFile = file.TempFile
```
Comment (hilmarf) [inline pkg/blobaccess/forward.go:99]: Why is this necessary to duplicate?

File: pkg/blobaccess/standard/access.go [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+func DataAccessForData(data []byte, origin ...string) bpi.DataSource {
+	path := ""
+	if len(origin) > 0 {
+		path = filepath.Join(origin...)
+	}
+	return &bytesAccess{data: data, origin: path}
+}
+
+func DataAccessForString(data string, origin ...string) bpi.DataSource {
```
Comment (hilmarf) [inline pkg/blobaccess/standard/access.go:33]:
```suggestion
func DataSourceForString(data string, origin ...string) bpi.DataSource {
```

**Review by hilmarf — CHANGES_REQUESTED** [review]
I would like to see more function documentation and maybe rename a few functions.

File: pkg/blobaccess/file/access.go (hunk 2 above)
Comment (mandelsoft) [reply pkg/blobaccess/file/access.go:17]: It makes is private for the embedding types. But here it is not really necessary,
, because it does not have fields.

File: pkg/blobaccess/file/access.go (hunk 5 above)
Comment (mandelsoft) [reply pkg/blobaccess/file/access.go:236]: I just aligned the package structure. The functions are still the same to be compatible.

Comment (mandelsoft) [reply pkg/blobaccess/file/access.go:236]: WithMeta means with additional meta data passes as arguments, here the mime type the digest and the size

File: pkg/blobaccess/file/access.go (hunk 6 above)
Comment (mandelsoft) [reply pkg/blobaccess/file/access.go:253]: The signature for these functions cannot be changed without being incompatible.

File: pkg/blobaccess/forward.go (hunk 10 above)
Comment (mandelsoft) [reply pkg/blobaccess/forward.go:33]: a DataSource is a specialized type for DataAccess

File: pkg/blobaccess/file/access.go (hunk 3 above)
Comment (mandelsoft) [reply pkg/blobaccess/file/access.go:58]: added

File: pkg/blobaccess/file/access.go (hunk 4 above)
Comment (mandelsoft) [reply pkg/blobaccess/file/access.go:190]: it also validates, but to do so, it has to request a lock first, here.

File: pkg/blobaccess/file/access.go (hunk 5 above)
Comment (mandelsoft) [reply pkg/blobaccess/file/access.go]: I added a new function with options, the old is deprecated.

File: pkg/blobaccess/file/access.go (hunk 6 above)
Comment (mandelsoft) [reply pkg/blobaccess/file/access.go]: same here

File: pkg/blobaccess/file/access.go (hunk 7 above)
Comment (mandelsoft) [reply pkg/blobaccess/file/access.go]: fixed

File: pkg/blobaccess/file/access.go (hunk 8 above)
Comment (mandelsoft) [reply pkg/blobaccess/file/access.go:298]: docu added

File: pkg/blobaccess/file/access.go (hunk 9 above)
Comment (mandelsoft) [reply pkg/blobaccess/file/access.go:327]: docu added

File: pkg/blobaccess/forward.go (hunk 11 above)
Comment (mandelsoft) [reply pkg/blobaccess/forward.go:42]: Blobaccess creation in a blockaccess package is for generally ForXXX. to avoid repetition of package name.
You can write now `blobaccess.ForXXX(...)`.

File: pkg/blobaccess/forward.go (hunk 12 above)
Comment (mandelsoft) [reply pkg/blobaccess/forward.go:46]: Same here `blobaccess.ProviderForXXX(...)`.

File: pkg/blobaccess/forward.go (hunk 13 above)
Comment (mandelsoft) [reply pkg/blobaccess/forward.go:51]: see above

File: pkg/blobaccess/forward.go (hunk 14 above)
Comment (mandelsoft) [reply pkg/blobaccess/forward.go:55]: see above

File: pkg/blobaccess/forward.go (hunk 15 above)
Comment (mandelsoft) [reply pkg/blobaccess/forward.go:95]: - for compatibility
- the new blobaccess package now provided access to all included blob access creation functions provided by sub packages. So you can wirite `blobaccess.ForXXX(...)`. the other central functions of implementation packages are forwarded, also.

File: pkg/blobaccess/standard/access.go (hunk 16 above)
Comment (mandelsoft) [reply pkg/blobaccess/standard/access.go:33]: see above

Comment (hilmarf) [reply pkg/blobaccess/standard/access.go]: `dataAccess.ForXXX(...)`?

File: pkg/blobaccess/blobaccess/doc.go [hunk 17]
```diff
@@ -0,0 +1,11 @@
+// Package blobaccess provides the basic set of types and supporting
+// functions for using BlobAccess implementations.
+// It is intended to be used by generic BlobAccess users not referring to
+// dedicated implementations.
+// Dedicated implementations can be accessed using the appropriate implementation
+// package. THis separation is required to avoid cycle for BlobAccess implementations
```
Comment (hilmarf) [inline pkg/blobaccess/blobaccess/doc.go:6]:
```suggestion
// package. This separation is required to avoid cycle for BlobAccess implementations
```

File: pkg/blobaccess/blobaccess/doc.go [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+// using again BlobAccess consumers.
+// Alternatively, the parent package can be used, if there are no cycles. The
+// parent package provide thes ebasic functionality plus
```
Comment (hilmarf) [inline pkg/blobaccess/blobaccess/doc.go:9]:
```suggestion
// parent package provide these basic functionality plus
```

File: pkg/blobaccess/blobaccess/doc.go [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+// access the to various implementation flavors in one package.
```
Comment (hilmarf) [inline pkg/blobaccess/blobaccess/doc.go:10]:
```suggestion
// access to the various implementation flavors in one package.
```

Comment (mandelsoft) [reply pkg/blobaccess/blobaccess/doc.go:10]: fixed

File: pkg/blobaccess/blobaccess/doc.go (hunk 18 above)
Comment (mandelsoft) [reply pkg/blobaccess/blobaccess/doc.go:9]: fixed

File: pkg/blobaccess/blobaccess/doc.go (hunk 17 above)
Comment (mandelsoft) [reply pkg/blobaccess/blobaccess/doc.go:6]: fixed

File: pkg/blobaccess/standard/access.go (hunk 16 above)
Comment (mandelsoft) [reply pkg/blobaccess/standard/access.go]: `dataAccess` is no package. It is contained in the `blobaccess` package. Therefore, there are two flavors:
- `blobaccess.ForXXX(...)`
- `blobaccess.DataAccessForXXX(...)` (optionally available)

File: pkg/blobaccess/forward.go (hunk 12 above)
Comment (mandelsoft) [reply pkg/blobaccess/forward.go:39]: similar to dataaccess this would lead to a bunch of additional packages. Therefore (as before) , there is only the `blobaccess`package. It contains
- the `DataAccess` base abstraction (plus the creator functions `DataAccesForXXX(...)
- the `BlobAccess`  abstraction extending the `DataAccess  (plus the creator functions `ForXXX(...)
- the `BlobAccessProvider` abstraction  (plus the creator functions `ProviderForXXX(...)

**Review by hilmarf — APPROVED** [review]
